{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='predict'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CellPy Predict\n",
    "\n",
    "## Predict Options\n",
    "\n",
    "predictOne: prediction of layers w/ val_metadata, each layer is predicted independently\n",
    "* (runMode = predictOne, predNormExpr, predMetadata, layerObjectPaths, rejectionCutoff)\n",
    "* Example: `cellpy --runMode predictOne --predNormExpr /cellpy_example/pbmc_10k_normalized.csv --predMetadata /cellpy_example/pbmc_10k_metadata.csv --layerObjectPaths /cellpy_example/cellpy_results_20210720155257/training/Root_object.pkl,/cellpy_example/cellpy_results_20210720155257/training/CD4_object.pkl,/cellpy_example/cellpy_results_20210720155257/training/CD8_object.pkl,/cellpy_example/cellpy_results_20210720155257/training/T-cell_object.pkl --rejectionCutoff 0.5`\n",
    "\n",
    "predictOne: prediction of layers w/o val_metadata, each layer is predicted independently\n",
    "* (runMode = predictOne, predNormExpr, layerObjectPaths, rejectionCutoff)\n",
    "* Example: `cellpy --runMode predictOne --predNormExpr /cellpy_example/pbmc_10k_normalized.csv --layerObjectPaths /cellpy_example/cellpy_results_20210720155257/training/Root_object.pkl,/cellpy_example/cellpy_results_20210720155257/training/CD4_object.pkl,/cellpy_example/cellpy_results_20210720155257/training/CD8_object.pkl,/cellpy_example/cellpy_results_20210720155257/training/T-cell_object.pkl --rejectionCutoff 0.5`\n",
    "\n",
    "predictAll: prediction of layers w/o val_metadata, each layer influences the next layer\n",
    "* (runMode = predictAll, predNormExpr, layerObjectPaths, rejectionCutoff)\n",
    "* Example: `cellpy --runMode predictAll --predNormExpr /cellpy_example/pbmc_10k_normalized.csv --layerObjectPaths /cellpy_example/cellpy_results_20210720155257/training/Root_object.pkl,/cellpy_example/cellpy_results_20210720155257/training/CD4_object.pkl,/cellpy_example/cellpy_results_20210720155257/training/CD8_object.pkl,/cellpy_example/cellpy_results_20210720155257/training/T-cell_object.pkl --rejectionCutoff 0.5`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## predictOne runMode Option (2a, 2b)\n",
    "\n",
    "predictOne is a CellPy option that allows the user to use the CellPy algorithm to predict cell types in an **individual layer** in their single cell RNA query dataset based off of their previously trained dataset.\n",
    "\n",
    "*Note: CellPy creates and stores information from model training in **Layer** objects, including the name of the layer, its depth in metadata files, the dictionary associated with each layer, and the XGBoost model itself. Prediction and feature ranking require **Layer** objects as input and can therefore only be run after **trainAll** has been completed. Details for the **Layer** class can be found at the end of this tutorial.*\n",
    "\n",
    "### REQUIRED USER INPUTS:\n",
    "-  **predNormExpr**: normalized expression matrix file\n",
    "-  **predMetadata (optional)**: metadata file\n",
    "-  **layerObjectPaths**: a list of path names to Layer objects\n",
    "    - **NOTE:** the option `cardiacDevAtlas` can be directly utilized as an argument to layerObjectPaths to predict cell types in your cardiac dataset based on the cardiac developmental atlas used in our paper\n",
    "-  **rejectionCutoff**: float between 0 and 1 denoting the minimum probability for a prediction to not be rejected\n",
    "\n",
    "### OUTPUT OF PREDICT:\n",
    "* creates directory 'prediction' in cellpy_results folder, defines 'Root' as topmost layer\n",
    "* csv files containing all the predictions and probabilities associated with each label for each cell\n",
    "* metric files detailing accuracy, precision, recall, confusion matrix (if metadata file has been provided in input; only present for predictOne mode)\n",
    "\n",
    "### predNormExpr\n",
    "\n",
    "##### FORMAT: csv file\n",
    "\n",
    "-  cell 0,0 / A1 is 'gene'\n",
    "-  first column contains gene names\n",
    "-  first row contains cell names\n",
    "-  eg. cell [i,j] is the gene expression value in cell (j-1) for gene (i-1)\n",
    "\n",
    "##### EXAMPLE:\n",
    "\n",
    "`gene,AAACCCAG........,AAACGAAC........,AAACGAAT........,GAGGGATC........\n",
    "MIR1302-10,0,2.14693417019908,2.31409562022533,0\n",
    "OR4F29,0,1.71783906814673,0,0\n",
    "LINC00115,0,0,0,2.8499342352407\n",
    "ISG15,2.99811039595896,0,2.41534932603235,0`\n",
    "\n",
    "### predMetadata (Optional)\n",
    "\n",
    "##### FORMAT: csv file\n",
    "\n",
    "-  cell 0,0 / A1 should be NA or empty\n",
    "-  first column contains cell names, should be in the same order as first row of trainNormExpr\n",
    "-  each column following contains the layered identification for each cell\n",
    "-  all other cells should be NA or empty\n",
    "\n",
    "The prediction metadata file option allows the user to test the model on a new dataset.\n",
    "EXAMPLE:\n",
    "\n",
    "`,Celltype1,Celltype2,Celltype3\n",
    "AAACCCAG........,Monocyte,NA,NA\n",
    "AAACGAAC........,NK,NA,NA\n",
    "AAACGAAT........,Macrophage,NA,NA\n",
    "GAGGGATC........,T-cell,\"CD8,\"Effector Memory CD8`\n",
    "\n",
    "### layerObjectPaths\n",
    "\n",
    "##### FORMAT: a comma-separated list of paths to the trained Layer objects (pickle .pkl files)\n",
    "\n",
    "* Layer objects were created by trainOnly\n",
    "-  not all models have to be provided, can conduct prediction on individual targetted layers\n",
    "* **NOTE: do not rename the .pkl Layer objects*\n",
    "\n",
    "* the cardiac developmental atlas datasets found in our paper can be utilized as a training dataset by simply using the **`cardiacDevAtlas`** option instead of inputting your own Layer objects\n",
    "\n",
    "##### EXAMPLES:\n",
    "\n",
    "\n",
    "training one layer:\n",
    "`--layerObjectPaths /scratch/groups/smwu/sidraxu/cellpy_results_20210720155257/training/Root_object.pkl`\n",
    "\n",
    "training multiple layers (predicted individually):\n",
    "`--layerObjectPaths /scratch/groups/smwu/sidraxu/cellpy_results_20210720155257/training/Root_object.pkl,/scratch/groups/smwu/sidraxu/cellpy_results_20210720155257/training/CD4_object.pkl,/scratch/groups/smwu/sidraxu/cellpy_results_20210720155257/training/CD8_object.pkl,/scratch/groups/smwu/sidraxu/cellpy_results_20210720155257/training/T-cell_object.pkl`\n",
    "\n",
    "### rejectionCutoff\n",
    "\n",
    "##### FORMAT: float between 0 and 1\n",
    "\n",
    "-  a rejection cutoff of 0.5 means a cell will be regarded as \"Unclassified\" if no class has a predicted probability greater than 50%\n",
    "* **NOTE: See \"Post-CellPy Analysis in R\" section below for further analysis on varying rejection thresholds' impact on results. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## predictAll runMode Option (3a)\n",
    "\n",
    "predictAll is a CellPy option that allows the user to use the CellPy algorithm to predict cell types in **all layers** of their single cell RNA query dataset based off of their previously trained dataset. Each layer of cell types predicted influences the predictions for the next layer.\n",
    "\n",
    "*Note: CellPy creates and stores information from model training in **Layer** objects, including the name of the layer, its depth in metadata files, the dictionary associated with each layer, and the XGBoost model itself. Prediction and feature ranking require **Layer** objects as input and can therefore only be run after **trainAll** has been completed. Details for the **Layer** class can be found at the end of this tutorial.*\n",
    "\n",
    "### REQUIRED USER INPUTS:\n",
    "-  **predNormExpr**: normalized expression matrix file\n",
    "-  **layerObjectPaths**: a list of path names to Layer objects\n",
    "    - **NOTE:** the option `cardiacDevAtlas` can be directly utilized as an argument to layerObjectPaths to predict cell types in your cardiac dataset based on the cardiac developmental atlas used in our paper\n",
    "-  **rejectionCutoff**: float between 0 and 1 denoting the minimum probability for a prediction to not be rejected\n",
    "\n",
    "### OUTPUT OF PREDICT:\n",
    "* creates directory 'prediction' in cellpy_results folder, defines 'Root' as topmost layer\n",
    "* csv files containing all the predictions and probabilities associated with each label for each cell\n",
    "* metric files detailing accuracy, precision, recall, confusion matrix (if metadata file has been provided in input; only present for predictOne mode)\n",
    "\n",
    "### predNormExpr\n",
    "\n",
    "##### FORMAT: csv file\n",
    "\n",
    "-  cell 0,0 / A1 is 'gene'\n",
    "-  first column contains gene names\n",
    "-  first row contains cell names\n",
    "-  eg. cell [i,j] is the gene expression value in cell (j-1) for gene (i-1)\n",
    "\n",
    "##### EXAMPLE:\n",
    "\n",
    "`gene,AAACCCAG........,AAACGAAC........,AAACGAAT........,GAGGGATC........\n",
    "MIR1302-10,0,2.14693417019908,2.31409562022533,0\n",
    "OR4F29,0,1.71783906814673,0,0\n",
    "LINC00115,0,0,0,2.8499342352407\n",
    "ISG15,2.99811039595896,0,2.41534932603235,0`\n",
    "\n",
    "### layerObjectPaths\n",
    "\n",
    "##### FORMAT: a comma-separated list of paths to the trained Layer objects (pickle .pkl files)\n",
    "\n",
    "* Layer objects were created by trainOnly\n",
    "-  not all models have to be provided, can conduct prediction on individual targetted layers\n",
    "* **NOTE: do not rename the .pkl Layer objects*\n",
    "\n",
    "* the cardiac developmental atlas datasets found in our paper can be utilized as a training dataset by simply using the **`cardiacDevAtlas`** option instead of inputting your own Layer objects\n",
    "\n",
    "##### EXAMPLE:\n",
    "\n",
    "`--layerObjectPaths /scratch/groups/smwu/sidraxu/cellpy_results_20210720155257/training/Root_object.pkl,/scratch/groups/smwu/sidraxu/cellpy_results_20210720155257/training/CD4_object.pkl,/scratch/groups/smwu/sidraxu/cellpy_results_20210720155257/training/CD8_object.pkl,/scratch/groups/smwu/sidraxu/cellpy_results_20210720155257/training/T-cell_object.pkl`\n",
    "\n",
    "### rejectionCutoff\n",
    "\n",
    "##### FORMAT: float between 0 and 1\n",
    "\n",
    "-  a rejection cutoff of 0.5 means a cell will be regarded as \"Unclassified\" if no class has a predicted probability greater than 50%\n",
    "* **NOTE: See \"Post-CellPy Analysis in R\" section below for further analysis on varying rejection thresholds' impact on results. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Back to Table of Contents\n",
    "\n",
    "[Table of Contents](tableofcontents.ipynb#toc)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
